First draft of Groovy 4 release notes / upgrade guide (#33633)
gradle/gradle: platforms/documentation/docs/src/docs/userguide/releases/upgrading/upgrading_version_8.adoc

NOTE: Plugins written using the Kotlin DSL and published with Gradle 7.x or 8.x remain compatible with Gradle 6.8 and newer.

==== Upgrade to Groovy 4.0.26

Groovy has been updated to https://groovy-lang.org/changelogs/changelog-4.0.26.html[4.0.26], from 3.0.24.
Also see the https://groovy-lang.org/releasenotes/groovy-4.0.html[4.0 release notes] for more details on the changes from 3.x to 4.x.

===== Properties with `is`-prefix and `Boolean` types are no longer usable by Groovy
Groovy 4 no longer detects getters with `is`-prefix and `Boolean` type as a property. Gradle's own interpretation of properties currently remains
the same, but will be changed in Gradle 10.0 to be consistent with Groovy 4. See <<#groovy_boolean_properties, the deprecation about this>> for more details.

> ===== Non-abstract managed properties may no longer be attached

Author: This isn't guaranteed, there are cases in which they get attached. "may" is the right word here.

**TODO, waiting on https://github.com/gradle/gradle/issues/33215 resolution**

===== Closures may prefer the delegate over the owner in more cases when using the `DELEGATE_FIRST` strategy
Groovy 4.0.26 has changed the behavior of closures using the `DELEGATE_FIRST` strategy. Dynamic lookups for properties and methods will now prefer the delegate over the owner.

> This can result in different behavior when using closures in Gradle scripts, such as certain methods not being found or (e.g. with `with(...) { }`) map keys being preferred over properties or methods

Reviewer: ```suggestion
This may lead to different behavior in Gradle scripts, for example, methods might not be found if shadowed by a map key (e.g., when using `with(...) { }`), as map entries now take precedence over outer properties or methods.
```

> from outside the closure.

Reviewer: ```suggestion
```

> Workarounds include using `@CompileStatic` to avoid the dynamic lookup, or prefixing the call with `owner.`, `this.` or `super.` as appropriate.

Reviewer: ```suggestion
Workarounds include using `@CompileStatic` to avoid dynamic lookups, or explicitly qualifying calls with `owner.`, `this.`, or `super.` as needed.
```

For full clarity, in Groovy 3, the lookup order was:

1. Delegate's `invokeMethod`, which chooses known methods and does no dynamic lookup.
2. Owner's `invokeMethod`, which chooses known methods and does no dynamic lookup.
3. Delegate's `invokeMissingMethod`, which does dynamic lookup including via properties.
4. Owner's `invokeMissingMethod`, which does dynamic lookup including via properties.

In Groovy 4, the lookup order is:

1. Delegate's `invokeMethod`, which chooses known methods and does no dynamic lookup.
2. Delegate's `invokeMissingMethod`, which does dynamic lookup including via properties.
3. Owner's `invokeMethod`, which chooses known methods and does no dynamic lookup.
4. Owner's `invokeMissingMethod`, which does dynamic lookup including via properties.

===== Private properties and methods may be inaccessible in closures

> Closures created in a parent class that reference properties or methods of the parent class may not be able to access them.

Reviewer: ```suggestion
Closures defined in a parent class that reference its private properties or methods may no longer have access to them.
```

> This is a https://issues.apache.org/jira/browse/GROOVY-11568[Groovy bug] that will not be fixed until Groovy 5.

Reviewer: ```suggestion
This is due to a https://issues.apache.org/jira/browse/GROOVY-11568[Groovy bug] that will not be resolved until Groovy 5.
```

> A workaround is to use `@CompileStatic` to avoid the dynamic lookup.

Reviewer: ```suggestion
As a workaround, apply `@CompileStatic` to bypass the dynamic lookup.
```

This is tracked by https://github.com/gradle/gradle/issues/32476.

===== Super methods may be inaccessible from Groovy 3.x code

> Groovy 4 changed how super methods are resolved, which may cause some super methods to be inaccessible from Groovy 3.x code,

Reviewer: ```suggestion
Groovy 4 changed how `super` method calls are resolved at runtime.
As a result, code compiled with Groovy 3.x may be unable to access `super` methods,
```

> as they do not contain the appropriate runtime code to locate the super methods.

Reviewer: ```suggestion
as it do not contain the appropriate runtime code to locate them.
```

> ==== Plugins written with the Groovy DSL require Gradle >= 7.0

Reviewer: Should this be a nested item?

```suggestion
===== Plugins written with the Groovy DSL require Gradle >= 7.0
```

Author: I left it as not nested, to match the Kotlin DSL entry.

> When building and publishing plugins using the Groovy DSL on Gradle 9.x, those plugins will only be usable on Gradle 7.0 or newer.

Reviewer: ```suggestion
Plugins built and published using the Groovy DSL on Gradle 9.x require Gradle 7.0 or newer to run.
```

> This is because Gradle 7.0 is the first release that embeds Groovy 3.0 or higher, which is required to partially run Groovy 4.0 compiled code.

Reviewer: What does it mean to "partially run code"?

> However, because it is still a major version change, some code may not work as expected.

Reviewer: ```suggestion
However, due to the major version change, some code may still behave unexpectedly.
```

> It is recommended to run these plugins on Gradle 9.0 or newer to ensure full compatibility with Groovy 4.0.

Reviewer: ```suggestion
For full compatibility, it's recommended to run these plugins on Gradle 9.0 or later.
```

NOTE: Plugins written using the Kotlin DSL and published with Gradle 7.x or 8.x remain compatible with Gradle 6.8 and newer.

==== Upgrade to Groovy 4.0.26

> Groovy has been upgraded from version 3.0.24 to https://groovy-lang.org/changelogs/changelog-4.0.26.html[4.0.26].

Author: I have never written them with that, so it should be fine. I think it's because they're unambiguous.

For a complete overview of changes from 3.x to 4.x, see the https://groovy-lang.org/releasenotes/groovy-4.0.html[Groovy 4.0 release notes].

===== `is`-prefixed `Boolean` properties no longer recognized by Groovy
Groovy 4 no longer treats getters with an `is` prefix and a `Boolean` return type as properties.
Gradle still recognizes these as properties for now, but this behavior will change in Gradle 10.0 to align with Groovy 4.
See <<#groovy_boolean_properties,the deprecation notice>> for more details.

===== Non-abstract managed properties may no longer be attached
**TODO, waiting on https://github.com/gradle/gradle/issues/33215 resolution**

> ===== `DELEGATE_FIRST` closures may now prefer the delegate in some cases

Author: I don't think it can affect any of our DSL, as we don't seem to have any APIs that have dynamic things in these types of closures. However, I could see someone using `with` on a `Jar`'s `Manifest.attributes`, i.e. `tasks.jar { with(manifest.attributes) { ... } }` and that could change things. I can add a script for that.

> ===== `DELEGATE_FIRST` closures may now prefer the delegate in some cases

Author: This doesn't actually change behavior since it's only on method invoke, and the attributes can't be invoked. I think the DSL should be pretty safe from this. But it triggered enough of our own test scripts that I think we should leave the warning here.

> ===== `DELEGATE_FIRST` closures may now prefer the delegate in some cases

Reviewer 2: can we say that we don't expect this to impact build scripts?

This can result in different behavior when using closures in Gradle scripts, such as certain methods not being found
or (e.g. with `with(...) { }`) some dynamic properties taking precedence over outer properties or methods.

Workarounds include using `@CompileStatic` to avoid dynamic lookups, or explicitly qualifying calls with `owner.`, `this.`, or `super.` as needed.

For full clarity, in Groovy 3, the lookup order was:

1. Delegate's `invokeMethod`, which chooses known methods and does no dynamic lookup.
2. Owner's `invokeMethod`, which chooses known methods and does no dynamic lookup.
3. Delegate's `invokeMissingMethod`, which does dynamic lookup including via properties.
4. Owner's `invokeMissingMethod`, which does dynamic lookup including via properties.

In Groovy 4, the lookup order is:

1. Delegate's `invokeMethod`, which chooses known methods and does no dynamic lookup.
2. Delegate's `invokeMissingMethod`, which does dynamic lookup including via properties.
3. Owner's `invokeMethod`, which chooses known methods and does no dynamic lookup.
4. Owner's `invokeMissingMethod`, which does dynamic lookup including via properties.

> ===== Private properties and methods may be inaccessible in closures

Author: Yes, it's the latter.

Closures defined in a parent class that reference its private properties or methods may no longer have access to them in subclasses.
This is due to a https://issues.apache.org/jira/browse/GROOVY-11568[Groovy bug] that will not be resolved until Groovy 5.
As a workaround, apply `@CompileStatic` to the class to remove the dynamic lookup.

This is tracked by https://github.com/gradle/gradle/issues/32476.

===== Super methods may be inaccessible from Groovy 3.x code
Groovy 4 changed how `super` method calls are resolved at runtime.
As a result, code compiled with Groovy 3.x may be unable to access `super` methods,
as the code does not contain the appropriate runtime code to locate them.

==== Plugins written with the Groovy DSL require Gradle >= 7.0

> Plugins built and published using the Groovy DSL on Gradle 9.x require Gradle 7.0 or newer to run.
> This is because Gradle 7.0 is the first release that embeds Groovy 3.0 or higher, which is required to run Groovy 4.0 compiled code.
> However, due to the major version change, some code may still behave unexpectedly.
> For full compatibility, it's recommended to run these plugins on Gradle 9.0 or later.
> 
> NOTE: Plugins written using the Groovy DSL and published with Gradle 7.x or 8.x remain compatible with Gradle 5.0 and newer.

Reviewer 2: ```suggestion
Plugins built and published using the Groovy DSL on Gradle 9.x are only supported on Gradle 7.0 or above. Gradle 7.0 is the first release that embeds Groovy 3.0 and can run Groovy 4.0 code. 

Since Gradle 9.0 uses Groovy 4.0, some code may behave unexpectedly. For full compatibility, it's recommended to run on Gradle 9.0 or later.

NOTE: Plugins written using the Groovy DSL and published with Gradle 7.x or 8.x remain compatible with Gradle 5.0 and newer.
```
